src/clj/rems/db/email_outbox.clj

```diff
@@ -0,0 +1,99 @@
+   :email-outbox/latest-attempt (:latest_attempt row)
+   :email-outbox/latest-error (:latest_error row)
+   :email-outbox/backoff (pg-interval->joda-duration (:backoff row))
+   :email-outbox/deadline (:deadline row)})
+
+(defn get-emails
+  ([]
+   (get-emails nil))
+  ([{:keys [ids due-now?]}]
+   (->> (db/get-email-outbox {:ids ids
+                              :due-now? due-now?})
+        (map fix-row-from-db))))
+
+(defn get-email-by-id [id]
+  (first (get-emails {:ids [id]})))
+
+(defn- next-attempt [email now error]
+  (let [^DateTime next-attempt (-> now (.plus (:email-outbox/backoff email)))
+        ^DateTime deadline (:email-outbox/deadline email)
+        ^Duration backoff (-> (:email-outbox/backoff email) (.multipliedBy 2))]
+    (assoc email
+           :email-outbox/latest-attempt now
+           :email-outbox/latest-error error
+           :email-outbox/next-attempt (if (-> next-attempt (.isAfter deadline))
+                                        nil
+                                        next-attempt)
+           :email-outbox/backoff (if (-> backoff (.isLongerThan max-backoff))
+                                   max-backoff
+                                   backoff))))
+
+(deftest test-next-attempt
+  (let [now (DateTime. 1000)]
+    (testing "basic case"
+      (is (= {:email-outbox/latest-attempt now
+              :email-outbox/latest-error "the error"
+              :email-outbox/next-attempt (DateTime. 3000) ; now + backoff
+              :email-outbox/backoff (Duration. 4000) ; 2 * backoff
+              :email-outbox/deadline (DateTime. 60000000)
+              :unrelated-keys "should be kept"}
+             (next-attempt {:email-outbox/backoff (Duration. 2000)
+                            :email-outbox/deadline (DateTime. 60000000)
+                            :unrelated-keys "should be kept"}
+                           now "the error"))))
+
+    (testing "max backoff reached"
+      (is (= {:email-outbox/latest-attempt now
+              :email-outbox/latest-error "the error"
+              :email-outbox/next-attempt (.plus now max-backoff)
+              :email-outbox/backoff max-backoff
+              :email-outbox/deadline (DateTime. 60000000)
+              :unrelated-keys "should be kept"}
+             (next-attempt {:email-outbox/backoff max-backoff
+                            :email-outbox/deadline (DateTime. 60000000)
+                            :unrelated-keys "should be kept"}
+                           now "the error"))))
+
+    (testing "deadline reached"
+      (is (= {:email-outbox/latest-attempt now
+              :email-outbox/latest-error "the error"
+              :email-outbox/next-attempt nil ; would have been 3000
+              :email-outbox/backoff (Duration. 4000)
+              :email-outbox/deadline (DateTime. 2999)
+              :unrelated-keys "should be kept"}
+             (next-attempt {:email-outbox/backoff (Duration. 2000)
```

Comment:
hmm this behaviour means that we might only retry for two days even though the max period is three days

it would perhaps be more predictable to always do the last retry precisely at the deadline?